More changes in AST fuzzer

PedroTadim opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

A few additions to the AST fuzzer: add join clauses to known tables, add optimize statements, and `DISTINCT`.
Added benchmark functionality when compiled with BuzzHouse. Using a peer ClickHouse server, compare the performance of each modified query (query time and memory usage). Sometimes settings are added to the query such as the join algorithm and block size.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory and will run independently of the checks below:
- [ ] <!---ci_include_stateless--> Only: Stateless tests
- [ ] <!---ci_include_integration--> Only: Integration tests
- [ ] <!---ci_include_performance--> Only: Performance tests
---
- [ ] <!---ci_exclude_style--> Skip: Style check
- [ ] <!---ci_exclude_fast--> Skip: Fast test
---
- [ ] <!---woolen_wolfdog--> Run all checks ignoring all possible failures (Resource-intensive. All test jobs execute in parallel).
- [ ] <!---no_ci_cache--> Disable CI cache

PedroTadim committed:
Add join clauses to the AST

PedroTadim committed:
Fuzz predicates and remove unused functions

PedroTadim committed:
Fix

PedroTadim committed:
Fuzz more settings

PedroTadim committed:
Negate predicates

PedroTadim committed:
Fuzz optimize commands and update probabilities

robot-clickhouse commented:
<!-- automatic status comment for PR #75279 from PedroTadim/ClickHouse:ast-update -->
*This is an automated comment for commit 2e8c62242a4a09c0e9432acc5da3ce34a3562f24 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/75279/2e8c62242a4a09c0e9432acc5da3ce34a3562f24/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d819a5d0e4c4f9b1fe582ec61689c9a053c045d7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

nikitamikhaylov commented on src/Common/QueryFuzzer.cpp:
```diff
@@ -1038,6 +1012,85 @@ struct ScopedIncrement
+                ASTPtr next_condition = nullptr;
+                const int nprob = fuzz_rand() % 10;
+
+                /// Pick a random indentifier
```
```suggestion
                /// Pick a random identifier
```

PedroTadim committed:
Update src/Common/QueryFuzzer.cpp

Co-authored-by: Nikita Mikhaylov <mikhaylovnikitka@gmail.com>

PedroTadim committed:
Don't fuzz children

PedroTadim committed:
MIssing checks

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Adding performance comparisons for AST fuzzer

PedroTadim committed:
I forgot to add fuzzed tables into OPTIMIZE statements

PedroTadim committed:
Always run query on peer

PedroTadim committed:
Merge branch 'ast-performance' into ast-update

PedroTadim committed:
Fix

PedroTadim committed:
Missing char

PedroTadim committed:
Compilation fix

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Missing check

maxknv committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into ast-update

PedroTadim committed:
Log performance results

PedroTadim committed:
Attempting multiple fixes

PedroTadim committed:
Fixes

PedroTadim committed:
Small fix

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Set random settings while measuring performance

maxknv committed:
Merge branch 'master' into ast-update

PedroTadim committed:
CI fix

PedroTadim committed:
Fixing join generation

PedroTadim committed:
Print seed used for AST fuzzer

PedroTadim committed:
Print set query

PedroTadim committed:
Fix

PedroTadim committed:
Set settings per query

PedroTadim committed:
Small fix

PedroTadim committed:
Fixes

PedroTadim committed:
Dump result into /dev/null

PedroTadim committed:
Fix

PedroTadim committed:
Code cleanup

PedroTadim committed:
Don't keep new settings

PedroTadim committed:
Fix

PedroTadim committed:
New metric

PedroTadim committed:
Fixes

PedroTadim committed:
Fix

PedroTadim committed:
Fixes

PedroTadim committed:
Another fix

PedroTadim committed:
Make sure settings are printed

PedroTadim committed:
More inner joins

PedroTadim committed:
Fuzz array joins

PedroTadim committed:
Fix

PedroTadim committed:
Fixes

PedroTadim committed:
Fix possible AST fuzzer bug

PedroTadim committed:
Compare query success and revert change

PedroTadim committed:
Dump fuzzed AST

PedroTadim committed:
Fixes

PedroTadim committed:
Setting too new

PedroTadim committed:
Fix

PedroTadim committed:
Fuzz join types

PedroTadim committed:
Fix

PedroTadim committed:
Dump results to /dev/null with INSERT INTO FUNCTION

PedroTadim committed:
Merge branch 'ClickHouse:master' into ast-update

PedroTadim committed:
Remove tables from FROM clause

PedroTadim committed:
More settings and load server settings at start

PedroTadim committed:
This step has to be done later

PedroTadim committed:
Don't AST fuzz added INSERT INTO /dev/null for performance testing

PedroTadim committed:
Make metric settings more generic

PedroTadim committed:
Fix tidy build

PedroTadim committed:
Add or remove distinct to aggregate

PedroTadim committed:
Fix

PedroTadim committed:
Merge branch 'master' into ast-update

pamarcos commented:
>I am planning to add benchmark checks next.

I think we should keep future plans out of the changelog entry. Of course, it's ok to add it to the PR description, but I find it very weird and even confusing to include future plans in a changelog.

PedroTadim commented:
> > I am planning to add benchmark checks next.
> 
> I think we should keep future plans out of the changelog entry. Of course, it's ok to add it to the PR description, but I find it very weird and even confusing to include future plans in a changelog.

The changelog was outdated. Updated it.

PedroTadim committed:
Updated settings for performance testing

PedroTadim committed:
Cleanup

PedroTadim committed:
Change aggregate functions and more settings for performance testing

PedroTadim committed:
Don't change cross joins

PedroTadim committed:
Add more aggregates to the list

PedroTadim committed:
Sometimes add to existing preficate

PedroTadim committed:
Swap more funcs and look for name convention

PedroTadim committed:
Check if old predicate exists

PedroTadim committed:
Fix clang-tidy build

PedroTadim committed:
Use settings more often

PedroTadim committed:
Permute expression lists

PedroTadim committed:
More interchangeable functions

PedroTadim committed:
More groups of  equivalent functions

PedroTadim committed:
Shuffle select from

PedroTadim committed:
Another shuffle

PedroTadim committed:
Add a few swappable functions and literals

PedroTadim committed:
Permute predicates

PedroTadim committed:
Fix tidy build

PedroTadim committed:
Look for or inside and and vice-versa

PedroTadim committed:
Permute join clauses

PedroTadim committed:
Use alias if the expression has one

PedroTadim committed:
My previous commit was not right

PedroTadim committed:
Clean a bit

PedroTadim committed:
Don't replace anything with a literal

PedroTadim committed:
Fuzz a few more strings

PedroTadim committed:
A and/or/xor function may have more than 2 arguments

PedroTadim committed:
Less asterisk

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Fix merge

PedroTadim committed:
Still fixing the merge

pamarcos commented:
@PedroTadim please add me back as a reviewer once you've finished pushing commits. I removed myself because I'm receiving a crazy amount of notifications and this seems to be still ongoing. I'd move it to draft if I were you

PedroTadim committed:
More groups of equivalent functions

PedroTadim committed:
Try to fuzz aliases better

PedroTadim committed:
Try to resuse alias elsewhere

PedroTadim committed:
Defensive lines

PedroTadim committed:
Group and order by from the projection

PedroTadim committed:
Fix

PedroTadim committed:
Add table functions to the table like list

PedroTadim committed:
push_back I think

PedroTadim committed:
Fix tody build

PedroTadim committed:
Fix offset

PedroTadim committed:
Fixes

PedroTadim committed:
Fix

PedroTadim committed:
Add array join and custom aliases to identifiers

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Fix tidy build

PedroTadim committed:
Try to reduce *

PedroTadim committed:
Set aliases at more contexts

PedroTadim committed:
Sometimes use random expressions and fix typo

PedroTadim committed:
Remove this branch, is not working properly

PedroTadim committed:
Less copying

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Remove include

PedroTadim committed:
Don't remove first FROM table

PedroTadim committed:
Fuzz casts and more data types

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Fix merge with master

PedroTadim committed:
Defensive line

PedroTadim commented on programs/client/Client.cpp:
```diff
@@ -838,12 +838,28 @@ void Client::processOptions(
         config().setBool("ignore-error", true);
         ignore_error = true;
+#if USE_BUZZHOUSE
+        if (query_fuzzer_runs && !buzz_house_options_path.empty())
```
The idea is to use BuzzHouse configuration file to tune the performance parameters

PedroTadim commented on programs/client/Client.h:
```diff
@@ -2,6 +2,9 @@
 #include <Client/ClientApplicationBase.h>
 
+#if USE_BUZZHOUSE
+#    include <Client/BuzzHouse/Generator/ExternalIntegrations.h>
```
I get this error: `unique_ptr.h:78:19: error: invalid application of 'sizeof' to an incomplete type 'BuzzHouse::FuzzConfig'` if I replace the header with a forward declaration.

pamarcos replied:
That's because the destructor of the class that holds the unique_ptr needs to be able to see the actual size of whatever unique_ptr points to in order to properly destroy it. So, the easiest solution for this is to define a default destructor in the .cpp, where you actually need to know the complete definition anyway.

PedroTadim replied:
Ok, but to avoid moving code, can we leave the include there? It doesn't slow the compilation that much

Algunenano replied:
NO. Don't include absolutely anything in headers if it's not strictly necessary. Otherwise I'll have to spent hours trying to gain back compilation time just to save you a few keystrokes.

pamarcos replied:
The problem are the transitive dependencies of header files, Pedro. Anything we can spare, we should take it. Especially for straight cases such as this one. It's just removing that line and adding a one-liner of default ctor in the .cpp

PedroTadim commented on programs/client/Client.h:
```diff
@@ -53,6 +56,9 @@ class Client : public ClientApplicationBase
     void printChangedSettings() const;
     void showWarnings();
 #if USE_BUZZHOUSE
+    std::unique_ptr<BuzzHouse::FuzzConfig> fc;
+    std::unique_ptr<BuzzHouse::ExternalIntegrations> ei;
```
Expanding the names. They are unique_ptrs because they are optional (if there is no BuzzHouse config). FuzzLoop.cpp uses them.

pamarcos replied:
But where are they actually used? Because at least VSCode (using clangd underneath) didn't find any references to those

PedroTadim replied:
Check FuzzLoop.cpp line 197

PedroTadim commented on programs/client/FuzzLoop.cpp:
```diff
@@ -7,8 +7,12 @@
 
 #include <Parsers/ASTCreateQuery.h>
 #include <Parsers/ASTDropQuery.h>
+#include <Parsers/ASTFunction.h>
```
Those includes are not mine, but I can try to fix them

PedroTadim committed:
Addressing review comments

PedroTadim committed:
Adress review changes

PedroTadim committed:
This header is still needed

PedroTadim committed:
More review comments

PedroTadim committed:
Arguments may be removed beforehand

pamarcos committed:
Remove unnecessary include

If no destructor is added, an implicit destructor is automatically
generated in the header file for every translation unit including
the header. Since it's the destructor the one that needs to know
the size of what std::unique_ptr points to in order to know how
to destroy it, the best thing we can do is to define a default
destructor in the .cpp to prevent any automatic destructors
being generated via header include.

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Merge branch 'master' into ast-update

pamarcos commented on src/Common/QueryFuzzer.cpp:
```diff
@@ -433,69 +484,60 @@ void QueryFuzzer::fuzzWindowFrame(ASTWindowDefinition & def)
     {
-        case 0:
-        {
+        case 0: {
```
Minor nitpick: Why did you move the start of the curly brace one line?
I think according to our coding style it should be on a separate line as it was before

PedroTadim replied:
I used clang-format and it didn't format this? :thinking:

PedroTadim replied:
I used clang-format in the whole file and it formated this. Am I doing something wrong?

pamarcos replied:
Probably not. Maybe it's a rule that is not set up properly in our clang-format file. It's not a big thing anyway
[Point 3 of our C++ coding style](https://clickhouse.com/docs/development/style#formatting)

Point 1 says:
> 1. Most of the formatting is done automatically by clang-format.

This is not one of those I guess :sweat:

pamarcos commented on src/Common/QueryFuzzer.cpp:
```diff
@@ -865,6 +937,35 @@ ASTs QueryFuzzer::getInsertQueriesForFuzzedTables(const String & full_query)
     return queries;
 }
 
+/// I am still not very expert in templates to reuse the code above
```
Kind reminder that you don't need to solve everything with templates. The two functions are extremely similar. You could do it with templates, sure. But even without them, you could have an extra flag parameter that uses the optimize query or the insert one.

PedroTadim replied:
I can try to clean this further

pamarcos commented on programs/client/FuzzLoop.cpp:
```diff
@@ -270,7 +386,7 @@ bool Client::processWithFuzzing(const String & full_query)
             have_error = false;
         }
-        else if (ast_to_process->formatForErrorMessage().size() > 500)
+        else if (ast_to_process->formatForErrorMessage().size() > 2000)
```
Why did you have to increase this value?

PedroTadim committed:
clang-format

PedroTadim committed:
Removing duplicated code

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Compilation fix

PedroTadim committed:
Fix error message

PedroTadim committed:
Fix peer check

PedroTadim committed:
Merge branch 'master' into ast-update

PedroTadim committed:
Compilation fixes

PedroTadim committed:
Merge branch 'ClickHouse:master' into ast-update

PedroTadim committed:
Merge remote-tracking branch 'origin/master' into ast-update

PedroTadim committed:
Add more performance settings and compilation fix

PedroTadim committed:
tidy build fix

PedroTadim committed:
Merge remote-tracking branch 'origin/master' into ast-update

Algunenano commented:
This is breaking the fuzzer itself instead of the server:
* https://s3.amazonaws.com/clickhouse-test-reports/json.html?REF=master&sha=3d01c4cf8e959e4098596a1b9ce30701eceaf433&name_0=MasterCI&name_1=AST%20fuzzer%20%28debug%29
* https://s3.amazonaws.com/clickhouse-test-reports/PRs/77713/e636a316fd5d220c67eaa400bdc37681c9e3e979//ast_fuzzer_debug/run.log


```
2025-03-17 10:26:48	 Thread 1 "clickhouse-clie" received signal SIGABRT, Aborted.
2025-03-17 10:26:48	 0x00007f5b611c29fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:48	 #0  0x00007f5b611c29fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:48	 No symbol table info available.
2025-03-17 10:26:48	 #1  0x00007f5b6116e476 in raise () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:48	 No symbol table info available.
2025-03-17 10:26:48	 #2  0x00007f5b611547f3 in abort () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:48	 No symbol table info available.
2025-03-17 10:26:48	 #3  0x00007f5b6115471b in ?? () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:48	 No symbol table info available.
2025-03-17 10:26:48	 #4  0x00007f5b61165e96 in __assert_fail () from /lib/x86_64-linux-gnu/libc.so.6
2025-03-17 10:26:49	 No symbol table info available.
2025-03-17 10:26:49	 #5  0x000055871dc73b2b in DB::ASTIdentifier::ASTIdentifier (this=0x7f5b5e8cf8d8, name_parts_=..., special=false, name_params=...) at /build/src/Parsers/ASTIdentifier.cpp:48
2025-03-17 10:26:49	         part = <optimized out>
2025-03-17 10:26:49	         __begin2 = <optimized out>
2025-03-17 10:26:49	         __range2 = <optimized out>
2025-03-17 10:26:49	         __end2 = <optimized out>
2025-03-17 10:26:49	 #6  0x0000558716125f73 in std::__1::construct_at[abi:se190107]<DB::ASTIdentifier, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, DB::ASTIdentifier*>(DB::ASTIdentifier*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&&) (__args=..., __location=<optimized out>) at /build/contrib/llvm-project/libcxx/include/__memory/construct_at.h:41
2025-03-17 10:26:49	 No locals.
2025-03-17 10:26:49	 #7  std::__1::__construct_at[abi:se190107]<DB::ASTIdentifier, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, DB::ASTIdentifier*>(DB::ASTIdentifier*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&&) (__args=..., __location=<optimized out>) at /build/contrib/llvm-project/libcxx/include/__memory/construct_at.h:49
2025-03-17 10:26:49	 No locals.
2025-03-17 10:26:49	 #8  _ZNSt3__116allocator_traitsINS_9allocatorIN2DB13ASTIdentifierEEEE9constructB8se190107IS3_JNS_6vectorINS_12basic_stringIcNS_11char_traitsIcEENS1_IcEEEENS1_ISC_EEEEEvTnNS_9enable_ifIXntsr15__has_constructIS4_PT_DpT0_EE5valueEiE4typeELi0EEEvRS4_SH_DpOSI_ (__args=..., __p=<optimized out>) at /build/contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:328
2025-03-17 10:26:49	 No locals.
2025-03-17 10:26:49	 #9  _ZNSt3__120__shared_ptr_emplaceIN2DB13ASTIdentifierENS_9allocatorIS2_EEEC2B8se190107IJNS_6vectorINS_12basic_stringIcNS_11char_traitsIcEENS3_IcEEEENS3_ISC_EEEEES4_TnNS_9enable_ifIXntsr7is_sameINT0_10value_typeENS_19__for_overwrite_tagEEE5valueEiE4typeELi0EEES4_DpOT_ (this=0x7f5b5e8cf8c0, __args=..., __a=...) at /build/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:264
2025-03-17 10:26:49	         __tmp = <optimized out>
2025-03-17 10:26:49	 #10 _ZNSt3__115allocate_sharedB8se190107IN2DB13ASTIdentifierENS_9allocatorIS2_EEJNS_6vectorINS_12basic_stringIcNS_11char_traitsIcEENS3_IcEEEENS3_ISA_EEEEETnNS_9enable_ifIXntsr8is_arrayIT_EE5valueEiE4typeELi0EEENS_10shared_ptrISE_EERKT0_DpOT1_ (__a=..., __args=...) at /build/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:843
2025-03-17 10:26:49	         __guard = {__alloc_ = {<std::__1::__non_trivial_if<true, std::__1::allocator<std::__1::__shared_ptr_emplace<DB::ASTIdentifier, std::__1::allocator<DB::ASTIdentifier> > > >> = {<No data fields>}, <No data fields>}, __n_ = 1, __ptr_ = 0x7f5b5e8cf8c0}
2025-03-17 10:26:49	         __control_block = <optimized out>
2025-03-17 10:26:49	 #11 0x000055871611c21a in _ZNSt3__111make_sharedB8se190107IN2DB13ASTIdentifierEJNS_6vectorINS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEENS7_IS9_EEEEETnNS_9enable_ifIXntsr8is_arrayIT_EE5valueEiE4typeELi0EEENS_10shared_ptrISD_EEDpOT0_ (__args=...) at /build/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:851
2025-03-17 10:26:49	 No locals.
2025-03-17 10:26:49	 #12 DB::QueryFuzzer::addJoinClause (this=this@entry=0x7ffffa366dc0) at /build/src/Common/QueryFuzzer.cpp:1446
2025-03-17 10:26:49	         id1_alias = {static __endian_factor = 1, __r_ = {<std::__1::__compressed_pair_elem<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__rep, 0, false>> = {__value_ = {__s = {__data_ = '\000' <repeats 22 times>, __padding_ = 0x7ffffa362ea7 "", __size_ = 0 '\000', __is_long_ = 0 '\000'}, __l = {__data_ = 0x0, __size_ = 0, __cap_ = 0, __is_long_ = 0}}}, <std::__1::__compressed_pair_elem<std::__1::allocator<char>, 1, true>> = {<std::__1::allocator<char>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<char> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, static npos = 18446744073709551615}
2025-03-17 10:26:49	         exp1 = {__ptr_ = 0x7ffffa365967, __cntrl_ = 0xc0}
2025-03-17 10:26:49	         rand_col1 = <optimized out>
2025-03-17 10:26:49	         nidentifier = <optimized out>
2025-03-17 10:26:49	         exp2 = {__ptr_ = 0x3237317361696c61, __cntrl_ = 0x0}
2025-03-17 10:26:49	         next_condition = {__ptr_ = 0x7f5b5e8bdd50, __cntrl_ = 0x7f5b5e8bdd80}
2025-03-17 10:26:49	         id1 = <optimized out>
2025-03-17 10:26:49	         to_search = <optimized out>
2025-03-17 10:26:49	         rand_col2 = {__i_ = 0x7f5b5e3f2f18}
2025-03-17 10:26:49	         i = 0
2025-03-17 10:26:49	         join_condition = {__ptr_ = 0x0, __cntrl_ = 0x0}
2025-03-17 10:26:49	         next_alias = {static __endian_factor = 1, __r_ = {<std::__1::__compressed_pair_elem<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__rep, 0, false>> = {__value_ = {__s = {__data_ = "alias172", '\000' <repeats 14 times>, __padding_ = 0x7ffffa362ec7 "\b\030\314?^[\177", __size_ = 8 '\b', __is_long_ = 0 '\000'}, __l = {__data_ = 0x3237317361696c61 <error: Cannot access memory at address 0x3237317361696c61>, __size_ = 0, __cap_ = 576460752303423488, __is_long_ = 0}}}, <std::__1::__compressed_pair_elem<std::__1::allocator<char>, 1, true>> = {<std::__1::allocator<char>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<char> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, static npos = 18446744073709551615}
2025-03-17 10:26:49	         rand_table = <optimized out>
2025-03-17 10:26:49	         input_table = <optimized out>
2025-03-17 10:26:49	         table_exp = {__ptr_ = 0x7f5b5e783958, __cntrl_ = 0x7f5b5e783940}
2025-03-17 10:26:49	         table_join = {__ptr_ = 0x7f5b5fbdbd18, __cntrl_ = 0x7f5b5fbdbd00}
2025-03-17 10:26:49	         old_alias = {static __endian_factor = 1, __r_ = {<std::__1::__compressed_pair_elem<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__rep, 0, false>> = {__value_ = {__s = {__data_ = '\000' <repeats 22 times>, __padding_ = 0x7ffffa362f0f "", __size_ = 0 '\000', __is_long_ = 0 '\000'}, __l = {__data_ = 0x0, __size_ = 0, __cap_ = 0, __is_long_ = 0}}}, <std::__1::__compressed_pair_elem<std::__1::allocator<char>, 1, true>> = {<std::__1::allocator<char>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<char> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, static npos = 18446744073709551615}
2025-03-17 10:26:49	         nconditions = <optimized out>
2025-03-17 10:26:49	         table = <optimized out>
```

Need to revert. Please revert the revert with the needed fixes

PedroTadim commented:
It's related to what I asked you about a few weeks back. I can add a workaround for this.
